src/main.rs

```diff
@@ -8,57 +8,86 @@ use std::path::PathBuf;
 
 #[derive(Parser, Debug)]
 #[command(author, version, about, long_about = None)]
-struct Args {
-  #[arg(value_enum)]
+struct Cli {
+  #[command(subcommand)]
   pub mode: Mode,
 
-  #[arg(short, long)]
+  #[arg(short, long, global = true)]
   pub verbose: bool,
+}
 
-  #[command(flatten)]
-  pub wopts: WOpts,
+#[derive(Subcommand, Clone, Debug)]
+enum Mode {
+  /// Checks that the program is syntactically and semantically correct.
+  Check {
+    #[arg(help = "Path to the input file")]
+    path: PathBuf,
+  },
+  /// Compiles the program to hvmc and prints to stdout.
+  Compile {
+    #[arg(
+      short = 'O',
+      value_delimiter = ' ',
+      action = clap::ArgAction::Append,
+      long_help = r#"Enables or disables the given optimizations
+      supercombinators is enabled by default."#,
+    )]
+    cli_opts: Vec<hvml::OptArgs>,
+
+    #[command(flatten)]
+    wopts: WOpts,
+
+    #[arg(help = "Path to the input file")]
+    path: PathBuf,
+  },
+  /// Compiles the program and runs it with the hvm.
+  Run {
+    #[arg(short, long, help = "How much memory to allocate for the runtime", default_value = "1G", value_parser = mem_parser)]
+    mem: usize,
 
-  #[arg(short, long, help = "Shows runtime stats and rewrite counts")]
-  pub stats: bool,
+    #[arg(short = 'd', help = "Debug mode (print each reduction step)")]
+    debug: bool,
 
-  #[arg(short, long, help = "How much memory to allocate for the runtime", default_value = "1G", value_parser = mem_parser)]
-  pub mem: usize,
+    #[arg(short = '1', help = "Single-core mode (no parallelism)")]
+    single_core: bool,
 
-  #[arg(short = '1', help = "Single-core mode (no parallelism)")]
-  pub single_core: bool,
+    #[arg(short = 'l', help = "Linear readback (show explicit dups)")]
+    linear: bool,
 
-  #[arg(short = 'd', help = "Debug mode (print each reduction step)")]
-  pub debug: bool,
+    #[arg(short, long = "stats", help = "Shows runtime stats and rewrite counts")]
+    arg_stats: bool,
 
-  #[arg(short = 'l', help = "Linear readback (show explicit dups)")]
-  pub linear: bool,
+    #[arg(help = "Path to the input file")]
+    path: PathBuf,
 
-  #[arg(
-    short = 'O',
-    value_delimiter = ' ',
-    action = clap::ArgAction::Append,
-    long_help = r#"Enables or disables the given optimizations
-    all, eta, no-eta, prune, no-prune, ref-to-ref, no-ref-to-ref,
-    supercombinators (enabled by default), no-supercombinators,
-    simplify-main, no-simplify-main, pre-reduce-refs, no-pre-reduce-refs
-    "#,
-  )]
-  pub opts: Vec<String>,
+    #[arg(
+      short = 'O',
+      value_delimiter = ' ',
+      action = clap::ArgAction::Append,
+      long_help = r#"Enables or disables the given optimizations
+      supercombinators is enabled by default."#,
+    )]
+    cli_opts: Vec<hvml::OptArgs>,
 
-  #[arg(help = "Path to the input file")]
-  pub path: PathBuf,
+    #[command(flatten)]
+    wopts: WOpts,
+  },
+  /// Runs the lambda-term level optimization passes.
```

Comment:
```suggestion
  /// Runs the lambda-term level desugaring passes.
```
Since we don't perform any optimizations here anymore